### PR TITLE
 [Draft] attach extra tensor to tensor builder 

### PR DESCRIPTION
**Pull request by zetwhite**
(WIP) This draft try to attach extra tensors to tensor builder.

ONE-DCO-1.0-Signed-off-by: SeungHui Youn <sseung.youn@samsung.com>

related : https://github.com/Samsung/ONE/issues/13282

### CODE REVIEW

**Comment by zetwhite** [comment]
[mnist_set.zip](https://github.com/user-attachments/files/16052998/mnist_set.zip)

**Comment by zetwhite** [comment]
@nnfw-bot test onert-cross-release

File: runtime/onert/core/src/compiler/ExecutorFactory.cc [hunk 1]
```diff
@@ -869,6 +875,16 @@ exec::IExecutor *ExecutorFactory::createTrainableExecutor(
...
+  // [idea2] : Generate Extra Tensors by traversing code_map
+
+  // Problem 1 :
+  // Since code_map lost its' backend context,
+  // it is hard to specify where estra tensor should be registered.
+
+  // Problme 2:
+  // Since ITrainableFunction is in exec area, it isn't a good idea to use member function of *Layer
+  // here, Istead of exec area interface, we have to implement another backend interface
+
```
Comment (zetwhite) [inline runtime/onert/core/src/compiler/ExecutorFactory.cc:887]: .

**Comment by hseok-oh** [comment]
IMO

- Introduce individual tensor manager (and memory manager if we need) for extra tensor
  - Extra tensor manager get request for extra tensor and manage extra tensor
  - Extra tensor manager may use extra memory manager to manage actual memory
- Train backend's BackendContext owns extra tensor manager, and train kernel generator points extra tensor manager  
- When kernel generator visits each operator, visitor request extra tensor manager(s)
  - Extra tensor manager return generated extra tensor: tensor struct with tensor info only, not allocated yet
  - Visitor create op's kernel layer and backward configure with extra tensor(s) from extra tensor manager
- After kernel generator's visitor loop end, kernel generator request extra tensor manager to allocate in `genKernels()` (or `generateFunctionMap()`)

**Comment by zetwhite** [comment]
@hseok-oh Thank you a lot for taking a look :smile: 

I catched your point. It looks quite good :+1: :smile:
 
Since I already implemented ExtraTensorManage and ExtraTensorIndex .. etc, I could quickly implement your suggestions. I'll revise it right now.

File: runtime/onert/backend/train/ExtraTensorIndex.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __ONERT_BACKEND_TRAIN_EXTRA_TENSOR_INDEX_H__
+#define __ONERT_BACKEND_TRAIN_EXTRA_TENSOR_INDEX_H__
+
+#include <cassert>
+#include <functional>
+#include <unordered_map>
```
[the remark this answers is not in the document]
Comment (zetwhite) [reply runtime/onert/backend/train/ExtraTensorIndex.h:22]: FYI, This draft is not ready for review :cry:  

So, this draft now have so many unnecessary changes now.

[the remark this answers is not in the document]
Comment (ragmani) [reply runtime/onert/backend/train/ExtraTensorIndex.h:22]: Sorry for my pointing out taking you by surprise, I just intended to just inform you the information.

[the remark this answers is not in the document]
Comment (zetwhite) [reply runtime/onert/backend/train/ExtraTensorIndex.h:22]: No problem at all. I really appreciate your interest :smile:

File: runtime/onert/backend/train/MemoryManager.cc [hunk 3]
```diff
@@ -53,52 +53,60 @@ uint8_t *GradientMemoryManager::getOptVarBuffer(const ir::OperandIndex &ind, uin
...
 
-uint8_t *DisposableMemoryManager::getBuffer(const DisposableTensorIndex &ind) const
+template <typename Index> uint8_t *TrainMemoryManager<Index>::getBuffer(const Index &ind) const
 {
   assert(_mem_planner->memory_plans().find(ind) != _mem_planner->memory_plans().end());
   const auto &mem_blk = _mem_planner->memory_plans().at(ind);
   return _mem_alloc->base() + mem_blk.offset;
 }
 
+// specialization
```
[the remark this answers is not in the document]
Comment (zetwhite) [reply runtime/onert/backend/train/MemoryManager.cc:106]: > If template member functions are defined in a cc file, can these instantiated template class be used in other cc files without including this cc file?

Yes, I don't know why. But this worked well.
I'd better check accurate C++ grammar related to this

[the remark this answers is not in the document]
Comment (ragmani) [reply runtime/onert/backend/train/MemoryManager.cc:106]: I'm very surprised that it works. I'm curious about the reason why there is no compilation problem.
However, I'm worried about multiple definition problem if those instantiated template classes are used in multiple cc files. Could you check that too?

[the remark this answers is not in the document]
Comment (zetwhite) [reply runtime/onert/backend/train/MemoryManager.cc:106]: >  I'm worried about multiple definition problem if those instantiated template classes are used in multiple cc files. Could you check that too?

That's why I put instantiation here. 
I first tried it in the header, but the compiler complains about multiple definitions. 

> I'm very surprised that it works. I'm curious about the reason why there is no compilation problem 

I couldn't fully understand why it works. But I could find similar answers in stack overflow : 
https://stackoverflow.com/questions/32707522/explicit-instantiation-of-template

[the remark this answers is not in the document]
Comment (ragmani) [reply runtime/onert/backend/train/MemoryManager.cc:106]: > I first tried it in the header, but the compiler complains about multiple definitions.

Usually, we have resolved by using `inline` keyword. Anyway, it seems that there is no complaint about multiple definition with the template instantiation.

> I couldn't fully understand why it works. But I could find similar answers in stack overflow :
https://stackoverflow.com/questions/32707522/explicit-instantiation-of-template

I also found a related answers in https://isocpp.org/wiki/faq/templates#separate-template-fn-defn-from-decl.

**Pull request closed without merging**